beacon-chain/sync/pending_blocks_queue.go

```diff
@@ -306,55 +306,120 @@ func (s *Service) sendBatchRootRequest(ctx context.Context, roots [][32]byte, ra
 	ctx, span := trace.StartSpan(ctx, "sendBatchRootRequest")
 	defer span.End()
 
+	// Exit early if there are no roots to request.
+	if len(roots) == 0 {
+		return nil
+	}
+
+	// Remove duplicates (if any) from the list of roots.
 	roots = dedupRoots(roots)
-	s.pendingQueueLock.RLock()
-	for i := len(roots) - 1; i >= 0; i-- {
-		r := roots[i]
-		if s.seenPendingBlocks[r] || s.cfg.chain.BlockBeingSynced(r) {
-			roots = append(roots[:i], roots[i+1:]...)
-		} else {
+
+	// Reversly iterate through the list of roots to request blocks, and filter out roots that are already
+	// seen in pending blocks or being synced.
+	func() {
+		s.pendingQueueLock.RLock()
+		defer s.pendingQueueLock.RUnlock()
+
+		for i := len(roots) - 1; i >= 0; i-- {
+			r := roots[i]
+			if s.seenPendingBlocks[r] || s.cfg.chain.BlockBeingSynced(r) {
+				roots = append(roots[:i], roots[i+1:]...)
+				continue
+			}
+
 			log.WithField("blockRoot", fmt.Sprintf("%#x", r)).Debug("Requesting block by root")
 		}
-	}
-	s.pendingQueueLock.RUnlock()
+	}()
 
+	// Nothing to do, exit early.
 	if len(roots) == 0 {
 		return nil
 	}
+
+	// Fetch best peers to request blocks from.
 	bestPeers := s.getBestPeers()
+
+	// Filter out peers that do not custody a superset of our columns.
```

Comment:
why ? we are requesting only blocks here not columns

Reply:
No. You can look in `sendRecentBeaconBlocksRequest` which is called by `sendBatchRootRequest`.